transaction/core/src/signer/traits.rs

```diff
@@ -0,0 +1,128 @@
+use crate::{
+    ring_signature::{CryptoRngCore, MLSAGError, ReducedTxOut, RingMLSAG, Scalar},
+    Amount,
+};
+use alloc::{string::String, vec::Vec};
+use displaydoc::Display;
+use mc_crypto_keys::{KeyError, RistrettoPrivate};
+use serde::{Deserialize, Serialize};
+use zeroize::Zeroize;
+
+/// A representation of the part of the input ring needed to create an MLSAG
+#[derive(Clone, Debug)]
+pub struct SignableInputRing {
+    /// A reduced representation of the TxOut's in the ring. For each ring
+    /// member we have only:
+    /// * The onetime-address (tx_out.target_key)
+    /// * The compressed commitment (tx_out.amount.commitment)
+    pub members: Vec<ReducedTxOut>,
+
+    /// The index of the real input among these ring members
+    pub real_input_index: usize,
+
+    /// The secrets needed to sign that input
+    pub input_secret: InputSecret,
+}
+
+/// The secrets needed to create a signature that spends an existing output as
+/// an input
+#[derive(Clone, Debug, Zeroize)]
+#[zeroize(drop)]
+pub struct InputSecret {
+    /// Represents either the one-time private key, or an alternative route to
+    /// this
+    pub onetime_key_or_alternative: OneTimeKeyOrAlternative,
+    /// The amount of the output
+    pub amount: Amount,
+    /// The blinding factor of the output we are trying to spend
+    pub blinding: Scalar,
+}
+
+/// To spend an input, we need to be able to derive the one-time private key.
+/// For off-line signing, we can't have this on the online machine. So in that
+/// case, we provide only the subaddress index, and the RingSigner (off-line
+/// machine) must use the account private keys to derive the one-time private
+/// key.
+///
+/// However, in e.g. the gift code flow, we must include the one-time
+/// private key from the gift code sender and we cannot possibly derive it
+/// ourselves.
+///
+/// This enum selects which path to the one-time private key is taken.
+#[derive(Clone, Debug, Zeroize)]
+#[zeroize(drop)]
+pub enum OneTimeKeyOrAlternative {
+    /// The one-time private key for the output
+    OneTimeKey(RistrettoPrivate),
+    /// The subaddress index which owns the output
+    SubaddressIndex(u64),
+}
```

Comment:
So the way I've currently set it up, this enum is just plain old data, and the `RingSigner` is responsible for getting the one-time key from this, since the ring signer has the private keys. So given that I don't really want to call it `OneTimeKeyDeriver`, since the other object is really the deriver. And I don't really want to add methods to this thing, I'd rather keep all the logic in the other thing.

I am not wedded to the `OneTimeKeyOrAlternative` name... 

How do you feel about `MaybeOneTimeKey` ?

Reply:
+1 for `OneTimeKeyInfo` or `OneTimeKeyMaterial`

Reply:
how about `OneTimeKeyDeriveData` ?